### PR TITLE
Feat/MET-4417_MET-4667 Records tier 0 backend

### PR DESCRIPTION
**Pull request by JoanaCMS**

### CODE REVIEW

File: src/main/java/eu/europeana/metis/sandbox/repository/RecordRepository.java [hunk 1]
```diff
@@ -47,4 +58,22 @@ public interface RecordRepository extends JpaRepository<RecordEntity, Long> {
...
       + "GROUP BY re.datasetId")
   List<DatasetStatistic> getMetricDatasetStatistics();
+
+  /**
+   * Find provider ids by dataset id and content tier value
+   *
+   * @param datasetId The id of the dataset the records belong to
+   * @param contentTier The value of the content tier to look for
+   */
+  @Query("select re.europeanaId from RecordEntity re where re.contentTier = ?2 and re.datasetId = ?1 order by re.europeanaId asc")
```
Comment (jeortizquan) [inline src/main/java/eu/europeana/metis/sandbox/repository/RecordRepository.java:68]: Something we noticed with Simon, to improve readability in SQL, is to use the keywords in Uppercase.
please update it with 
```SELECT re.europeanaId FROM RecordEntity re WHERE re.datasetId = ?1 AND re.contentTier = ?2 ORDER BY re.europeanaId ASC```
also the order of fields matter here in the sense when parsing the query, it will use the new index you specified on ```record``` table

File: src/main/java/eu/europeana/metis/sandbox/repository/RecordRepository.java [hunk 2]
```diff
@@ -47,4 +58,22 @@ public interface RecordRepository extends JpaRepository<RecordEntity, Long> {
...
+  List<String> findByDatasetIdAndContentTier(String datasetId, String contentTier);
+
+  /**
+   * Find provider ids by dataset id and content tier value
+   *
+   * @param datasetId The id of the dataset the records belong to
+   * @param metadataTier The value of the metadata tier to look for
+   */
+  @Query("select re.europeanaId from RecordEntity re where re.metadataTier = ?2 and re.datasetId = ?1 order by re.europeanaId asc")
```
Comment (jeortizquan) [inline src/main/java/eu/europeana/metis/sandbox/repository/RecordRepository.java:77]: Same here...
```SELECT re.europeanaId FROM RecordEntity re WHERE  re.datasetId = ?1 AND re.metadataTier = ?2 ORDER BY re.europeanaId ASC```

File: src/main/resources/database/schema.sql [hunk 3]
```diff
@@ -67,5 +69,6 @@ CREATE TABLE IF NOT EXISTS default_transform_xslt
 CREATE INDEX ON record_log (record_id);
 CREATE INDEX ON record_error_log (record_id);
 CREATE INDEX ON record (dataset_id, europeana_id, provider_id);
+CREATE INDEX ON record (dataset_id, content_tier, metadata_tier);
```
Comment (jeortizquan) [inline src/main/resources/database/schema.sql:72]: according to the new queries  in place of
```
CREATE INDEX ON record (dataset_id, content_tier, metadata_tier);
```
my suggestion would be
```
CREATE INDEX ON record (dataset_id, content_tier, europeana_id) 
CREATE INDEX ON record (dataset_id, metadata_tier, europeana_id)
```
what do you think?

File: src/test/java/eu/europeana/metis/sandbox/controller/DatasetControllerTest.java [hunk 4]
```diff
@@ -29,9 +29,7 @@
 import eu.europeana.metis.sandbox.common.locale.Country;
 import eu.europeana.metis.sandbox.common.locale.Language;
 import eu.europeana.metis.sandbox.dto.DatasetInfoDto;
-import eu.europeana.metis.sandbox.dto.report.ErrorInfoDto;
-import eu.europeana.metis.sandbox.dto.report.ProgressByStepDto;
-import eu.europeana.metis.sandbox.dto.report.ProgressInfoDto;
+import eu.europeana.metis.sandbox.dto.report.*;
```
Comment (jeortizquan) [inline src/test/java/eu/europeana/metis/sandbox/controller/DatasetControllerTest.java:32]: Please check this.
Is better to have the specific includes and avoid ``.*`` includes
```
import eu.europeana.metis.sandbox.dto.report.ErrorInfoDto;
import eu.europeana.metis.sandbox.dto.report.ProgressByStepDto;
import eu.europeana.metis.sandbox.dto.report.ProgressInfoDto;
```

Comment (jeortizquan) [reply src/test/java/eu/europeana/metis/sandbox/controller/DatasetControllerTest.java:32]:
```
import eu.europeana.metis.sandbox.dto.report.ErrorInfoDto;
import eu.europeana.metis.sandbox.dto.report.ProgressByStepDto;
import eu.europeana.metis.sandbox.dto.report.ProgressInfoDto;
import eu.europeana.metis.sandbox.dto.report.TierStatistics;
import eu.europeana.metis.sandbox.dto.report.TiersZeroInfo;
```

File: src/main/java/eu/europeana/metis/sandbox/service/dataset/DatasetReportServiceImpl.java [hunk 5]
```diff
@@ -7,14 +7,14 @@
...
+import eu.europeana.indexing.tiers.model.MetadataTier;
 import eu.europeana.metis.sandbox.common.Status;
 import eu.europeana.metis.sandbox.common.Step;
 import eu.europeana.metis.sandbox.common.exception.InvalidDatasetException;
 import eu.europeana.metis.sandbox.common.exception.ServiceException;
 import eu.europeana.metis.sandbox.dto.DatasetInfoDto;
-import eu.europeana.metis.sandbox.dto.report.ErrorInfoDto;
-import eu.europeana.metis.sandbox.dto.report.ProgressByStepDto;
-import eu.europeana.metis.sandbox.dto.report.ProgressInfoDto;
+import eu.europeana.metis.sandbox.dto.report.*;
```
Comment (jeortizquan) [inline src/main/java/eu/europeana/metis/sandbox/service/dataset/DatasetReportServiceImpl.java:17]: have a look here is preferrable to avoid ```.*``` and to use specific imports.
```
import eu.europeana.metis.sandbox.dto.report.ErrorInfoDto;
import eu.europeana.metis.sandbox.dto.report.ProgressByStepDto;
import eu.europeana.metis.sandbox.dto.report.ProgressInfoDto;
import eu.europeana.metis.sandbox.dto.report.TierStatistics;
import eu.europeana.metis.sandbox.dto.report.TiersZeroInfo;
```

File: src/main/java/eu/europeana/metis/sandbox/service/dataset/DatasetReportServiceImpl.java [hunk 6]
```diff
@@ -116,10 +121,14 @@ public ProgressInfoDto getReport(String datasetId) {
         recordsProcessedByStep.forEach((step, statusMap) -> addStepInfo(stepsInfo, statusMap, step,
                 recordErrorsByStep));
 
+        // prepare object TiersZeroInfo
```
Comment (jeortizquan) [inline src/main/java/eu/europeana/metis/sandbox/service/dataset/DatasetReportServiceImpl.java:124]: comment is not necessary, the method already is explaning what is doing.

File: src/main/java/eu/europeana/metis/sandbox/service/workflow/IndexingServiceImpl.java [hunk 7]
```diff
@@ -4,36 +4,50 @@
...
 
   @Override
   public RecordInfo index(Record recordToIndex) {
     requireNonNull(recordToIndex, "Record must not be null");
+    TierResults tierCalculations;
 
     try {
-      publishIndexer.index(recordToIndex.getContentInputStream(),
+      tierCalculations = publishIndexer.indexAndGetTierCalculations(recordToIndex.getContentInputStream(),
               new IndexingProperties(new Date(), false, null, false, true));
```
Comment (jeortizquan) [inline src/main/java/eu/europeana/metis/sandbox/service/workflow/IndexingServiceImpl.java:39]: This is out of scope of this ticket I just noticed ```IndexingProperties```. Please add a TODO: here please, to many different boolean parameters indicates many posibilities to create this object. This has to be refactored in a way that code is readable.

File: src/test/java/eu/europeana/metis/sandbox/dto/report/ProgressInfoDtoTest.java [hunk 8]
```diff
@@ -158,4 +158,4 @@ void getErrorType() {
     assertEquals( "Error", progressInfoDto.getErrorType());
     assertEquals( "", progressInfoDto.getPortalPublishUrl());
   }
-}
\ No newline at end of file
+}
```
Comment (jeortizquan) [inline src/test/java/eu/europeana/metis/sandbox/dto/report/ProgressInfoDtoTest.java:161]: You can include one more unit test here
To cover TierZeroInfo Object when is not null and check what is expected.

**Review by jeortizquan — CHANGES_REQUESTED** [review]
I see good job here integrating this new feature, please have a look on the comments.

File: src/main/java/eu/europeana/metis/sandbox/dto/report/TiersZeroInfo.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+ */
+public class TiersZeroInfo {
+
+    @JsonProperty("content-tier")
+    @JsonInclude(JsonInclude.Include.NON_NULL)
+    private TierStatistics contentTier;
+    @JsonProperty("metadata-tier")
+    @JsonInclude(JsonInclude.Include.NON_NULL)
+    private TierStatistics metadataTier;
+
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply src/main/java/eu/europeana/metis/sandbox/dto/report/TiersZeroInfo.java:20]: Done

File: src/main/java/eu/europeana/metis/sandbox/dto/report/TierStatistics.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    public int getTotalNumberOfRecords() {
+        return totalNumberOfRecords;
+    }
+
+    public List<String> getListRecordIds() {
+        return listRecordIds.isEmpty() ? Collections.emptyList() : Collections.unmodifiableList(listRecordIds);
+    }
+
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply src/main/java/eu/europeana/metis/sandbox/dto/report/TierStatistics.java:37]: Thanks for pointing it out! I just added it

File: src/main/java/eu/europeana/metis/sandbox/dto/report/TiersZeroInfo.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    public TierStatistics getContentTier() {
+        return contentTier;
+    }
+
+    public TierStatistics getMetadataTier() {
+        return metadataTier;
+    }
+
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply src/main/java/eu/europeana/metis/sandbox/dto/report/TiersZeroInfo.java:38]: Added

File: src/main/java/eu/europeana/metis/sandbox/repository/RecordRepository.java (hunk 1 above)
Comment (JoanaCMS) [reply src/main/java/eu/europeana/metis/sandbox/repository/RecordRepository.java:68]: Fixed it

File: src/main/resources/database/schema.sql (hunk 3 above)
Comment (JoanaCMS) [reply src/main/resources/database/schema.sql:72]: That is better indeed, I will put this suggestion

File: src/main/java/eu/europeana/metis/sandbox/service/dataset/DatasetReportServiceImpl.java (hunk 5 above)
Comment (JoanaCMS) [reply src/main/java/eu/europeana/metis/sandbox/service/dataset/DatasetReportServiceImpl.java:17]: The `*` happens automatically so I don't always notice this :')
Fixed it

File: src/main/java/eu/europeana/metis/sandbox/service/dataset/DatasetReportServiceImpl.java (hunk 6 above)
Comment (JoanaCMS) [reply src/main/java/eu/europeana/metis/sandbox/service/dataset/DatasetReportServiceImpl.java:124]: Removed

File: src/main/java/eu/europeana/metis/sandbox/service/workflow/IndexingServiceImpl.java (hunk 7 above)
Comment (JoanaCMS) [reply src/main/java/eu/europeana/metis/sandbox/service/workflow/IndexingServiceImpl.java:39]: Added

File: src/test/java/eu/europeana/metis/sandbox/controller/DatasetControllerTest.java (hunk 4 above)
Comment (JoanaCMS) [reply src/test/java/eu/europeana/metis/sandbox/controller/DatasetControllerTest.java:32]: Fixed

File: src/test/java/eu/europeana/metis/sandbox/dto/report/ProgressInfoDtoTest.java (hunk 8 above)
Comment (JoanaCMS) [reply src/test/java/eu/europeana/metis/sandbox/dto/report/ProgressInfoDtoTest.java:161]: Added

File: src/main/java/eu/europeana/metis/sandbox/dto/report/TierStatistics.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+import java.util.List;
+
+/**
+ * Class to encapsulate the statistics information related to a tier
+ *
+ */
+public class TierStatistics {
+
+    @JsonProperty("total")
+    private int totalNumberOfRecords;
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply src/main/java/eu/europeana/metis/sandbox/dto/report/TierStatistics.java]: Done

File: src/main/java/eu/europeana/metis/sandbox/dto/report/TierStatistics.java [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+
+        TierStatistics other = (TierStatistics) o;
+
+        return this.totalNumberOfRecords == other.totalNumberOfRecords &&
+                this.listRecordIds.equals(other.listRecordIds);
+    }
+
+    @Override
+    public int hashCode(){
+        return totalNumberOfRecords + listRecordIds.hashCode();
```
Comment (jochen-vermeulen) [inline src/main/java/eu/europeana/metis/sandbox/dto/report/TierStatistics.java:55]: Same as below. Use 'Objects.hash()'.

File: src/main/java/eu/europeana/metis/sandbox/dto/report/TiersZeroInfo.java [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+        TiersZeroInfo other = (TiersZeroInfo) o;
+
+        return this.contentTier.equals(other.contentTier) &&
+                this.metadataTier.equals(other.metadataTier);
+    }
+
+    @Override
+    public int hashCode(){
+        return contentTier.hashCode() + metadataTier.hashCode();
```
Comment (jochen-vermeulen) [inline src/main/java/eu/europeana/metis/sandbox/dto/report/TiersZeroInfo.java:55]: It's best to use a more complex function than addition to combine two hash functions. Luckily, Java has a neat solution:
`Objects.hash(contentTier, metadataTier);`

File: src/main/java/eu/europeana/metis/sandbox/service/dataset/DatasetReportServiceImpl.java [hunk 15]
```diff
@@ -238,9 +250,29 @@ private List<ErrorInfoDto> addStepErrors(Map<Status, Map<String, List<ErrorLogVi
...
-
         return Stream.of(recordEntity.getEuropeanaId(), recordEntity.getProviderId()).filter(
                 Objects::nonNull).filter(id -> !id.isBlank()).collect(Collectors.joining(" | "));
     }
 
+    private TiersZeroInfo prepareTiersInfo(String datasetId){
+        // get list of records with content tier 0
+        List<String> listOfRecordsIdsWithContentZero = recordRepository.findByDatasetIdAndContentTier(datasetId, MediaTier.T0.toString());
+        // get list of records with metadata tier 0
+        List<String> listOfRecordsIdsWithMetadataZero = recordRepository.findByDatasetIdAndMetadataTier(datasetId, MetadataTier.T0.toString());
```
Comment (jochen-vermeulen) [inline src/main/java/eu/europeana/metis/sandbox/service/dataset/DatasetReportServiceImpl.java:261]: There should be a way to limit the querying to just 10 records, instead of getting all records and then just taking the first 10? Seems like overkill (and a waste of IO).

File: src/main/resources/database/schema.sql (hunk 3 above)
Comment (jochen-vermeulen) [reply src/main/resources/database/schema.sql]: It could make sense perhaps as you're not just retrieving the `europeana_id`, but also first sorting by it. So perhaps there is some benefit? I don't know, though. So I suppose you could leave it like this.

File: src/main/java/eu/europeana/metis/sandbox/dto/report/TiersZeroInfo.java (hunk 14 above)
Comment (JoanaCMS) [reply src/main/java/eu/europeana/metis/sandbox/dto/report/TiersZeroInfo.java:55]: Thanks! This is indeed a better solution, I will change this

File: src/main/java/eu/europeana/metis/sandbox/dto/report/TierStatistics.java (hunk 13 above)
Comment (JoanaCMS) [reply src/main/java/eu/europeana/metis/sandbox/dto/report/TierStatistics.java:55]: Done

File: src/main/java/eu/europeana/metis/sandbox/service/dataset/DatasetReportServiceImpl.java (hunk 15 above)
Comment (JoanaCMS) [reply src/main/java/eu/europeana/metis/sandbox/service/dataset/DatasetReportServiceImpl.java:261]: Why I gathered everything first and then get the top 10 was because I wanted to get the count of the total records as well.
But that is a good point, I have created new methods in the record repository to gather the top 10 records as well as counting the number of records
